ft-sdk/src/in_.rs

```diff
@@ -1,19 +1,77 @@
-pub struct In {
+use std::cell::RefCell;
+use std::collections::HashMap;
+
+pub struct In<'c> {
     pub ud: Option<ft_sdk::UserData>,
     pub req: http::Request<bytes::Bytes>,
     pub now: chrono::DateTime<chrono::Utc>,
-    pub set_cookies: std::collections::HashMap<String, String>,
-    pub form_errors: std::collections::HashMap<String, String>,
+    pub set_cookies: RefCell<Vec<cookie::Cookie<'c>>>,
```

Comment:
Not fond of this. Can't we use native types (let's not expose dependency on other crates as part of our signature). Owned values for cookies is good enough, this optimisation means a lot of code has to deal with lifetimes now. Not great for end users.

Reply:
> Create a struct `Cookie` and use that, add other fields needed (domain, expiry, path etc) as option strings etc on it.

the `cookie` crate provides a very nice builder that I would personally want to use to create cookies. That said, if you really don't like this lifetime, we can do a `Vec<String>` and call `cookie::Cookie::to_string()` before storing them in the `In::set_cookies`

Reply:
Let's keep the string as the interface. If someone wants they can use cookie crate as a dependency. 

Also we do store In object on our own Layout implementation types.

Reply:
> Let's keep the string as the interface. If someone wants they can use cookie crate as a dependency.

Okay, I'll use `Vec<String>` for `set_cookies`. And, add a feature flag to expose the `In::add_cookie()` method. But, this will still not allow them to store the `In` object without dealing with lifetimes, here's why:

```rust
        // code copied from ft_sdk::Layout::_page(). similar code is for _action()
        let in_ = ft_sdk::In::from_request(r)?;
        let mut l = Self::from_in(&in_, RequestType::Page)?;
        let p = P::page(&in_, &mut l)?;
        let vj = serde_json::to_value(&p).unwrap();
        let oj = l.json(vj)?;
        Ok(ft_sdk::json_response(oj, Some(&in_)))
```

In order to get the updated cookies back from the wasm app, I'll have give them a reference to the `In` I own.

> Also we do store In object on our own Layout implementation types.

yes. But, do we have to if we have a reference to `In` in `page()`, and `action()`?

Reply:
RefCell is cheap to clone and each clone access the same data. 

I would avoid adding cookie even without a feature flag. As it ties us to another crate. Let's add a helper for common case, storing cookie with name value and expiry (domain and path most people don't want to update).

Reply:
> ... each clone access the same data.

I don't think this is true:

```rust
use std::cell::RefCell;

struct Foo {
    cookies: RefCell<Vec<String>>,
}

fn main() {
    let x = Foo {
        cookies: RefCell::new(Vec::new()),
    };

    x.cookies.borrow_mut().push("hello".to_string());

    // not a ref but a clone
    smth(x.cookies.clone());

    dbg!(x.cookies.borrow());
}

fn smth(x: RefCell<Vec<String>>) {
    x.borrow_mut().push("world".to_string());
    dbg!(x.borrow());
}
```

Output:

```
[I] sidd@codes ~/p/w/f/foo (feat/cookies)> cargo run 
    Finished dev [unoptimized + debuginfo] target(s) in 0.08s 
     Running `target/debug/fo` 
[src/main.rs:21:5] x.borrow() = [ 
    "hello", 
    "world", 
] 
[src/main.rs:16:5] x.cookies.borrow() = [ 
    "hello", 
] 
```

As you can see the Vec inside `main` and `smth` are different! `smth` gets the set cookies, but I don't get the cookies set by `smth` in main. So, unless we give smth a reference to `x`, we can't get changes made by `smth` in `main`.
We can go with clones and not change the param of `Layout::from_in()` but this will require us to change the return signature of `page()`, and `action()` and basically any functions that we would like to set cookies from. I don't like this.


> I would avoid adding cookie even without a feature flag. As it ties us to another crate. Let's add a helper for common > case, storing cookie with name value and expiry (domain and path most people don't want to update).

I agree with this. we can decide the default cookie for wasm apps giving them limited control (only over the name and value and expiry).

Reply:
https://doc.rust-lang.org/std/rc/struct.Rc.html#impl-Clone-for-Rc%3CT,+A%3E

You have to use `Rc<RefCell<T>>` pattern actually for `Clone`.